### PR TITLE
Netbox SAML integration

### PR DESCRIPTION
**Pull request by javierdelapuente**
### Overview

This PR adds SAML to NetBox using the SAML integration.

Besides the saml integration, It adds the next config variables to configure SAML:
- saml_sp_entity_id
- saml_username
- saml_email

For the integration tests it uses the [saml-test-idp](https://github.com/canonical/saml-test-idp.git) library.

There are several things that will be improved in next PRs, as Django 12 factor will make them work. This includes how to handle the saml env variables in case of error, that at the moment is not done, and will be done with the new `is_ready` function.

### Rationale

Add saml authentication and user registration to NetBox.

### Juju Events Changes

New events for SAML. Also, the ingress events now run the reconcile function.

### Module Changes

Changes to the env variables, so the needed configuration in NetBox can be applied.

### Library Changes

Added saml library.

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Manging Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [x] The documentation for charmhub is updated.
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

### CODE REVIEW

File: charm/charmcraft.yaml [hunk 1]
```diff
@@ -57,6 +61,20 @@ config:
...
+    saml_username:
+      description: |-
+        Saml attribute used for both the social uid and the username.
+        NameID can be used setting this config to 'name_id'.
+      type: string
+      default: 'name_id'
+    saml_email:
+      description: |-
+        Saml attribute used for the email.
+      type: string
```
Comment (jdkandersson) [inline charm/charmcraft.yaml:77]: Are these not provided by the SAML integration?

File: charm/src/charm.py [hunk 2]
```diff
@@ -34,8 +41,73 @@ def gen_env(self) -> dict[str, str]:
...
 
+    def saml_env(self) -> dict[str, str]:
+        """Environment variables for SAML.
+
+        Returns:
+           dict with environment variables.
+        """
+        saml_data: dict[str, str] = {}
+        if (relation := self.model.get_relation(self._SAML_RELATION_NAME)) is None:
+            return saml_data
```
Comment (jdkandersson) [inline charm/src/charm.py:61]: Couldn't these just return `{}`?

File: charm/src/charm.py [hunk 3]
```diff
@@ -34,8 +41,73 @@ def gen_env(self) -> dict[str, str]:
...
+        # https://github.com/canonical/charm-relation-interfaces/tree/main/interfaces/saml/v0
+        entity_id = relation_data.get("entity_id", "")
+        metadata_url = relation_data.get("metadata_url", "")
+        single_sign_on_service_redirect_url = relation_data.get(
+            "single_sign_on_service_redirect_url", ""
+        )
+        x509certs = relation_data.get("x509certs", "")
+
+        # saml_env should not raise. Pending to update this code once
+        # the next Django 12 factpr PR factor PR is merged, and maybe
```
Comment (jdkandersson) [inline charm/src/charm.py:80]: `factor`?

File: charm/src/charm.py (hunk 2 above)
Comment (javierdelapuente) [reply charm/src/charm.py]: Sure! Thanks for the advice

**Review by mthaddon — APPROVED** [review]
LGTM, thx!

**Pull request merged**